pkg/security/security_profile/activity_tree/activity_tree.go

```diff
@@ -598,27 +599,29 @@ func (at *ActivityTree) findProcessCacheEntryInTree(tree []*ProcessNode, entry *
 // findProcessCacheEntryInChildExecedNodes look for entry in the execed nodes of child
 func (at *ActivityTree) findProcessCacheEntryInChildExecedNodes(child *ProcessNode, entry *model.ProcessCacheEntry) *ProcessNode {
 	// children is used to iterate over the tree below child
-	execChildren := []*ProcessNode{child}
+	execChildren := make([]*ProcessNode, 1, 64)
+	execChildren[0] = child
+
+	visited := make([]*ProcessNode, 0, 64)
 
 	for len(execChildren) > 0 {
-		cursor := execChildren[0]
-		execChildren = execChildren[1:]
+		cursor := execChildren[len(execChildren)-1]
+		execChildren = execChildren[:len(execChildren)-1]
+
+		visited = append(visited, cursor)
 
 		// look for an execed child
 		for _, node := range cursor.Children {
-			if node.Process.IsExecChild {
+			if node.Process.IsExecChild && !slices.Contains(visited, node) {
 				// there should always be only one
-				execChildren = append(execChildren, node)
-			}
-		}
 
-		if len(execChildren) == 0 {
-			break
-		}
+				// does this execed child match the entry ?
+				if node.Matches(&entry.Process, at.differentiateArgs) {
```

Comment:
match check in line with iteration (skip iteration if not needed, no alloc if a match is found)